toxic_slicer.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"math/rand"
+	"time"
+)
+
+// The SlicerToxic slices data into multiple smaller packets
+// to simulate real-world TCP behaviour.
+type SlicerToxic struct {
+	Enabled bool `json:"enabled"`
+	// Average number of bytes to slice at
+	AverageSize int `json:"average_size"`
+	// +/- bytes to vary sliced amounts. Must be less than
+	// the average size
+	SizeVariation int `json:"size_variation"`
+	// Microseconds to delay each packet. May be useful since there's
+	// usually some kind of buffering of network data
+	Delay int `json:"delay"`
+}
+
+func (t *SlicerToxic) Name() string {
+	return "slicer"
+}
+
+func (t *SlicerToxic) IsEnabled() bool {
+	return t.Enabled
+}
+
+func (t *SlicerToxic) SetEnabled(enabled bool) {
+	t.Enabled = enabled
+}
+
+// Returns a list of chunk offsets to slice up a packet of the
+// given total size. For example, for a size of 100, output might be:
+//
+//     []int{0, 18, 18, 43, 43, 67, 67, 77, 77, 100}
+//           ^---^  ^----^  ^----^  ^----^  ^-----^
+//
+// This tries to get fairly evenly-varying chunks (no tendency
+// to have a small/large chunk at the start/end).
+func (t *SlicerToxic) chunk(start int, end int) []int {
+	// Base case:
+	// If the size is within the random varation, _or already
+	// less than the average size_, just return it.
+	// Otherwise split the chunk in about two, and recurse.
+	if (end-start)-t.AverageSize <= t.SizeVariation {
+		return []int{start, end}
+	}
+
+	// +1 in the size variation to offset favoring of smaller
+	// numbers by integer division
+	mid := start + (end-start)/2 + (rand.Intn(t.SizeVariation*2) - t.SizeVariation) + rand.Intn(1)
+	left := t.chunk(start, mid)
+	right := t.chunk(mid, end)
+
+	out := make([]int, 0, len(left)+len(right))
+	out = append(out, left...)
+	out = append(out, right...)
```

Comment:
Just did a quick benchmark since you got me curious.
The memory allocation is the same unless `len(right) < len(left)`, in which case using `make()` will use a bit less.
In almost all cases the size of `left` and `right` will be almost exactly the same (depending on `SizeVariation`), so I'm going to say I'd prefer shorter code of marginally faster code.

If performance is an issue with this, then the bigger target would be to not do recursive heap allocations.